<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">

<title>SmartMap</title>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<link rel="stylesheet" href="./index_files/style.css">
</head>

<body>

<div class="container">
	<!-- Definition of the sidebar. -->
	<div class="sidebar1">
		<h1>Multi<span style="color: #bf1300">Map</span></h1>
		<a href="#intro">Introduction</a>
		<button class="dropdown-btn"> 
			<i class="fa fa-caret-down"></i>SmartMapPrep
		</button>
		<div class="dropdown-container">
			<a href="#multimapprep">Description</a>
			<a href="#mmp-dep">Dependencies</a>
			<a href="#mmp-man">Manual</a>
		</div>
		<button class="dropdown-btn"> 
			<i class="fa fa-caret-down"></i>SmartMapRNAPrep
		</button>
		<div class="dropdown-container">
			<a href="#multimaprnaprep">Description</a>
			<a href="#mmrp-dep">Dependencies</a>
			<a href="#mmrp-man">Manual</a>
		</div>
		<button class="dropdown-btn"> 
			<i class="fa fa-caret-down"></i>SmartMap
		</button>
		<div class="dropdown-container">
			<a href="#multimap">Description</a>
			<a href="#mm-dep">Dependencies</a>
			<a href="#mm-install">Install</a>
			<a href="#mm-man">Manual</a>
		</div>
		<a href="#contact">Contact</a>
		<img src="./index_files/logo_big.png" alt="RuthenLab Logo">
	</div>

	<!-- Page-Specific Content. -->

	<div class="content">
		<h1>Bayesian Analysis of Ambiguously Mapped Reads</h1>
		<hr>
		<div class="bannerholder">
		  <img src="./index_files/home_banner.jpg">
		</div>
		<hr>
		<h2 id = "intro"><h2>Introduction</h2>
		<p>This tool is designed to analyze ambiguously mapped reads from paired-end short-read next-generation sequencing by using the alignment scores and the 
		distribution of mapped reads genome-wide to iteratively reweight and refine the weights assigned to each alignment of each read. This tool can be used 
		for both strand-nonspecific and strand-specific applications. In principle, single-end data can also be analyzed with the SmartMap tool, but the 
		SmartMapPrep and SmartMapRNAPrep scripts do not support single-end analysis, though modification of these scripts to use single-end data is possible if 
		desired.</p>
		
		<p>There are three components to the tool presented here: the SmartMapPrep and SmartMapRNAPrep bash scripts and the SmartMap compiled binary. 
		The SmartMapPrep and SmartMapRNAPrep scripts serves to align the FASTQ files to the genome, filter the output file, and parse it into a BED file 
		that can be used with the SmartMap binary, with and without strand information, respectively. The SmartMap binary is a compiled program written in 
		C++ to analyze the alignment BED file produced by SmartMapPrep or SmartMapRNAPrep with the iterative Bayesian algorithm and produce a BEDGRAPH file 
		of alignment weights across the genome, which can be treated as the weighted genome coverage BEDGRAPH file (with strand-specificity, if applicable).</p>
		
		<p>This tool is designed for use with Bowtie2 or Hisat2.</p>

		<hr>
		
		<h2 id = "multimapprep"><h2>Documentation for SmartMapPrep</h2>
		<h3 id = "mmp-desc"><h3>Description</h3>
		<p>The SmartMapPrep script is used to align the FastQ files and prepare a BED file output that is ready to be processed using the SmartMap software. This is accomplished by:</p>
		<ol>
			<li>Aligning the FastQ files to the genome using Bowtie2 using the following settings:
				<ul>
					<li>End-to-end alignment with very fast preset</li>
					<li>No discordant alignments</li>
					<li>No mixed alignments (single-end alignments if paired-end alignment cannot be found)</li>
					<li>Report up to <em>k</em> alignments per read pair (default 51)</li>
					<li>Insert size between <em>I</em> to <em>L</em> base pairs (default 100-250 bp)</li>
				</ul>
			</li>
			<li>Filtering for reads with one of the following SAM flags:
				<ul>
					<li>99 (read paired, mapped in a proper pair, mate reverse strand, first in pair, primary alignment)</li>
					<li>163 (read paired, mapped in a proper pair, mate reverse strand, second in pair, primary alignment)</li>
					<li>355 (read paired, mapped in a proper pair, mate reverse strand, first in pair, not primary alignment)</li>
					<li>419 (read paired, mapped in a proper pair, mate reverse strand, second in pair, not primary alignment)</li>
				</ul>
			</li>
			<li>Extracting into an extended BED file:
				<ul>
					<li>Chromosome</li>
					<li>Start position</li>
					<li>End position</li>
					<li>Read name</li>
					<li>Read alignment score (AS:i:)</li>
					<li>Mate alignment score (YS:i:)</li>
				</ul>
			</li>
			<li>Split the reads into separate files based on the number of alignments per read.</li>
		</ol>
		<p>The output files are a Gzipped file containing all alignments and a directory <code>splits</code> with unzipped files containing extended BED files 
		of the alignments split by number of alignments per read, prepared for line counting or use with the SmartMap software to run the iterative reweight 
		algorithm.</p>

		<h3 id="mmp-dep"><h3>Dependencies</h3>
		<p>In addition to standard Unix tools, including awk, sed, and gzip, the SmartMapPrep script requires the following to be installed and added to the PATH environment variable.</p>
		<ul>
			<li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a></li>
		</ul>
		<p>We have tested this script on Ubuntu LTS 14.04, 16.04, and 18.04. For the above listed tools, we have tested Bowtie2 v. 2.3.4.1.</p>
		
		<br>
		
		<h3 id="mmp-man">Manual for SmartMapPrep</h3>
		<div class="codebox">
			<p>SmartMapPrep [options] -x [Bowtie2 index] -o [output prefix] -1 [R1 fastq] -2 [R2 fastq]</p>
		</div>
		
		<p>Inputs (required):</p>
		<div class="codebox">
			<p>-x : Path to basename of Bowtie2 index for alignment</p>
			<p>-o : Output prefix prepended to the output files</p>
			<p>-1 : FastQ file for read mate 1 (can be gzipped)</p>  
			<p>-2 : FastQ file for read mate 2 (can be gzipped)</p>
		</div>
		
		<p>Options:</p>
		<div class="codebox">
			<p>-p : Number of CPU threads to be used for multithreaded alignment (default: 1)</p>
			<p>-I : Minimum insert length (default: 100)</p>
			<p>-L : Maximum insert length (default: 250)</p>
			<p>-k : Maximum number of alignments reported (default: 51)</p>
			<p>-s : String to be removed from read names</p>
			<br>
			<p>-h : Display help message</p>
		</div>
		
		<hr>
		
		<h2 id="multimaprnaprep">Documentation for SmartMapRNAPrep</h2>
		<h3 id="mmrp-desc">Description</h3>
		
		<p>The SmartMapRNAPrep script is used to align the FastQ files and prepare a BED file output that is ready to be processed using the SmartMap software. This is accomplished by:</p>
		<ol>
			<li>Aligning the FastQ files to the genome using Hisat2 using the following settings:
				<ul>
					<li>No discordant alignments</li>
					<li>No mixed alignments (single-end alignments if paired-end alignment cannot be found)</li>
					<li>Report up to <em>k</em> alignments per read pair (default 51)</li>
				</ul>
			</li>
			<li>Filtering for reads with one of the following SAM flags:
				<ul>
					<li>99 (read paired, mapped in a proper pair, mate reverse strand, first in pair, primary alignment)</li>
					<li>163 (read paired, mapped in a proper pair, mate reverse strand, second in pair, primary alignment)</li>
					<li>355 (read paired, mapped in a proper pair, mate reverse strand, first in pair, not primary alignment)</li>
					<li>419 (read paired, mapped in a proper pair, mate reverse strand, second in pair, not primary alignment)</li>
				</ul>
			</li>
			<li>Extracting into an extended BED file:
				<ul>
					<li>Chromosome</li>
					<li>Start position</li>
					<li>End position</li>
					<li>Read name</li>
					<li>Strand</li>
					<li>Read alignment score (AS:i:)</li>
					<li>Mate alignment score (YS:i:)</li>
				</ul>
			</li>
			<li>Split the reads into separate files based on the number of alignments per read.</li>
		</ol>
		<p>The output files are a Gzipped file containing all alignments and a directory <code>splits</code> with unzipped files containing extended BED files 
		of the alignments split by number of alignments per read, prepared for line counting or use with the SmartMap software to run the iterative reweight 
		algorithm.</p>
		
		<h3 id="mmrp-dep">Dependencies</h3>
		<p>In addition to standard Unix tools, including awk, sed, and gzip, the SmartMapPrep script requires the following to be installed and added to the PATH environment variable.</p>
		<ul>
		<li><a href="https://daehwankimlab.github.io/hisat2/">Hisat2</a></li>
		</ul>
		<p>We have tested this script on Ubuntu LTS 14.04, 16.04, and 18.04. For the above listed tools, we have tested Hisat2 v. 2.1.0.</p>
		
		<h3 id="mmrp-man">Manual for SmartMapRNAPrep</h3>
		<div class="codebox">
			<p>SmartMapRNAPrep [options] -x [Hisat2 index] -o [output prefix] -1 [R1 fastq] -2 [R2 fastq]</p>
		</div>
		
		<p>Inputs (required):</p>
		<div class="codebox">
			<p>-x : Path to basename of Hisat2 index for alignment</p>
			<p>-o : Output prefix prepended to the output files</p>
			<p>-1 : FastQ file for read mate 1 (can be gzipped)</p>
			<p>-2 : FastQ file for read mate 2 (can be gzipped)</p>
		</div>
		
		<p>Options:</p>
		<div class="codebox">
			<p>-p : Number of CPU threads to be used for multithreaded alignment (default: 1)</p>
			<p>-k : Maximum number of alignments reported (default: 51)</p>
			<p>-s : String to be removed from read names</p>
			<br>
			<p>-h : Display help message</p>
		</div>
		
		<hr>
		
		<h2 id="multimap">Documentation for SmartMap</h2>
		<h3 id="mm-desc">Description</h3>
		
		<p>The SmartMap software serves to assign weights to each mapping of each read by an iterative Bayesian reweighting algorithm using the BED file(s) 
		outputted by the SmartMapPrep script as an input. The output file is a Gzipped BEDGRAPH file of the genome-wide read weights analagous to a genome 
		coverage read depth BEDGRAPH file. If run in strand-specific mode, the output file is a pair of Gzipped BEDGRAPH files of the genome-wide read weights, 
		with one file per strand.</p>
		
		<h3 id="mm-dep">Dependencies</h3>
		<p>The SmartMap binary has no dependencies in and of itself; however, building the SmartMap binary from source requires the following to be installed and added to the PATH environment variable.</p>
		<ul>
		<li>g++</li>
		</ul>
		
		<p>Additionally, the <a href="https://www.cs.unc.edu/Research/compgeom/gzstream/">gzstream</a> library is required for building the SmartMap binary; this is included in the appropriate relative 
		directory with the appropriate reference in the makefile.<p>
		
		<p>We have tested this software on Ubuntu LTS 14.04, 16.04, and 18.04. For the above listed tools, we have tested g++ v. 7.5.0.</p>
		
		<h3 id="mm-install">Building SmartMap Binary from Source</h3>
		
		<p>The source files for the SmartMap software are provided here. To build the SmartMap binary from source, first navigate to the "Default" directory. Once in the "Default" directory, run the following command:</p>
		
		<div class="codebox">
			<p>make all</p>
		</div>
		
		<p>The SmartMap binary will appear in the "Default" directory. It can then be moved to the directory of interest or added to the PATH environment variable for use.</p>
		
		<h3 id="mm-man">Manual for SmartMap</h3>
		
		<div class="codebox">
			<p>SmartMap [options] -g [genome length file] -o [output prefix] [BED or BED.gz file input(s)]</p>
		</div>
		
		<p>Inputs (required):</p>
		<div class="codebox">
			<p>-g : Genome length file listing all chromosomes and lengths. Chromosomes will appear in this order in the output BEDGRAPH file.</p>
			<p>-o : Output prefix used for output BEDGRAPH and log files.</p>
			<p>Input files: Path to BED or BED.gz files, separated by spaces. Must be the last argument(s) passed to the software.</p>
		</div>
		
		<p>Options:</p>
		<div class="codebox">
			<p>-i : Number of iterations. Default 1.</p>
			<p>-x : Fixation percentage threshold. Default 0.</p>
			<p>-m : Maximum number of alignments for a read to be processed. Default 50.</p>
			<p>-s : Minimum score for Bowtie2 display. Default 0 (unscored).</p>
			<p>-v : N for N-fold cross-validation. Default 1 (no cross-validation).</p>
			<p>-c : Flag for continuous output bedgraphs. Default off.</p>
			<p>-S : Flag for strand-specific mode. Default off.</p>
			<br>
			<p>-h : Display help message</p>
		</div>
		
		<hr>
		
		<h2 id="contact">Contact and Acknowlegements</h2>
		<p>Contact Rohan Shah (rohanshah@uchicago.edu) or Alex Ruthenburg (aruthenburg@uchicago.edu) with questions, comments, or issues.</p>
		<p>Code and scripts for SmartMap is provided at the <a href="https://www.github.com/shah-rohan/SmartMap">SmartMap</a> GitHub repository.
		These tools, site, and associated GitHub repository were designed, written, and maintained by Rohan Shah.</p>
		<p>This Software is provided under a modified MIT License. Please cite and acknowledge this repository in any works utilizing this software.</p>
	</div>
</div>

<script>
/* Loop through all dropdown buttons to toggle between hiding and showing its dropdown content - This allows the user to have multiple dropdowns without any conflict */
var dropdown = document.getElementsByClassName("dropdown-btn");
var i;

for (i = 0; i < dropdown.length; i++) {
  dropdown[i].addEventListener("click", function() {
  this.classList.toggle("active");
  var dropdownContent = this.nextElementSibling;
  if (dropdownContent.style.display === "block") {
  dropdownContent.style.display = "none";
  } else {
  dropdownContent.style.display = "block";
  }
  });
}
</script>

</body></html>
